### ISSUE TITLE
Bring your own multimodal model?

### ISSUE THREAD

**Issue by frickp** [zou-group/textgrad#94]
Thanks for this really interesting project, it looks promising and I'd like to see if it will work for my multimodal prompt optimization. I see in this [code](https://github.com/zou-group/textgrad/blob/4a5063e4aff8b85e64d61a476c72de7c066f7ed7/textgrad/engine/__init__.py#L13L20) that there is a pre-defined set of models that can be multimodal and this blocks my usage. I am using Gemini, but generally it could be more flexible to allow users to bring _any_ multimodal model. I wonder if you think it would work to register an [engine](https://github.com/zou-group/textgrad/blob/4a5063e4aff8b85e64d61a476c72de7c066f7ed7/textgrad/engine/base.py#L5) with a `self.multimodal` attribute or something?

**Comment by mertyg** [comment]
Yes this would be great to have. I am not sure what is the best way to let users bring any multimodal model though -- for text, we can just use APIs that can be served through OpenAI clients [like this](https://github.com/zou-group/textgrad/blob/main/examples/notebooks/Local-Model-With-LMStudio.ipynb); but with images I am not sure if there is an easy way. Do you have any suggestions?

**Comment by frickp** [comment]
I was able to do it with Gemini Pro by subclassing the engine and overwriting the "generate" method. This is more overhead for the user, but allows it to work with arbitrary API designs across vendors. Not sure if you have a plan to loop in the huggingface ecosystem of models, but that may be relevant to this as well. [Here](https://github.com/zou-group/textgrad/blob/ee13ed76c57685d168453012bc9b8333df5c0547/textgrad/autograd/multimodal_ops.py#L69) it looks like if you can get it to generate with a list of inputs and a system prompt it should work for multimodal optimization. 

I can do it with Gemini like this:

```python
class GeminiMM(EngineLM, CachedEngine):
    def __init__(
        self,
        gemini_model: GenerativeModel, # This a model class from Google
        generation_config, # something Gemini expects, could be refactored
        safety_settings, # something Gemini expects, could be refactored
        model_string='gemini-vertex',
        system_instruction="You are a financial assistant."""
    ):
        self.model = gemini_model
        self.generation_config = generation_config # something Gemini expects, could be refactored
        self.safety_settings = safety_settings # something Gemini expects, could be refactored
        self.model_string = model_string
        

    @retry(wait=wait_random_exponential(min=1, max=5), stop=stop_after_attempt(5))
    def __call__(self, prompt, system_prompt, **kwargs):
        return self.generate(prompt, system_prompt, **kwargs)

    def generate(
        self,
        contents: list[Union[str, Part]], # this can accept text, image, in any order
        system_prompt,
    ):
        self.model._system_instruction = system_prompt
        generation = self.model.generate_content(
            contents,
            generation_config=self.generation_config,
            safety_settings=self.safety_settings,
        )
        return generation.text

```

If the validation check was in `EngineLM` I think I could get around the error.

**Comment by vinid** [comment]
yea in theory you should be able to use multimodal models with litellm!
